src/net_processing.cpp

```diff
@@ -655,6 +655,27 @@ class PeerManagerImpl final : public PeerManager
     bool AlreadyHaveTx(const GenTxid& gtxid)
         EXCLUSIVE_LOCKS_REQUIRED(cs_main, !m_recent_confirmed_transactions_mutex);
 
+    /**
+     * Map of seen chain tip sets keyed by unique network ID.
+     *
+     * The size of each chain tip set is limited to the number of chain tips in
+     * our global block index.
+     */
+    std::map<uint64_t, std::set<uint256>> m_chain_tip_sets GUARDED_BY(cs_main);
```

Comment:
Looks like this data structure is append-only for each new connection?

Reply:
It is append only (just like the block index) but not every new connection will add something to it. If a connection from a new network is made, then a new `std::set<uint256>` is added to the map for that network. The `std::set<uint256>` itself is limited to the number of tips in the block index.

Reply:
I had that same worry when writing this code but OOM errors should not be possible unless there is a bug in my implementation.

The logic for adding an entry to the `m_chain_tip_sets` map is the same logic that is used for the addr cache, so it can have at most as many entries as the addr cache (ie one for each network/bound address).

The chain tips sets are only modified through `ChainstateManager::UpdateChainTipSet` and an entry can only be added to the set if it already exists in the block index. See: https://github.com/dergoegge/bitcoin/blob/0f816ff2e60c5e440543ebb738db29bd285facb0/src/validation.cpp#L5259-L5260
If the new entry has one of the tips in the tip set as an ancestor than the existing tip is replaced, so there will never be more tips in each tip set than there are tips in the global index.

Reply:
> Looks like this data structure is append-only for each new connection?

@MarcoFalke note that this map is keyed by the unique network id, i.e. a combination of the peer's network and the local bound address/port. For most nodes, there is only one unique network id (ipv4 with a single bound local address), but for nodes running tor/i2p/etc or with multiple -bind addresses, there may be a handful.